src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2019-2023 Alibaba Cloud
+// Copyright (c) 2019-2023 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{collections::HashMap, sync::Arc};
+
+use anyhow::{anyhow, Context, Result};
+use kata_sys_util::rand::RandomBytes;
+use tokio::sync::Mutex;
+
+use crate::{
+    BlockConfig, Device, DeviceConfig, Hypervisor, KATA_BLK_DEV_TYPE, KATA_MMIO_BLK_DEV_TYPE,
+    VIRTIO_BLOCK_MMIO, VIRTIO_BLOCK_PCI,
+};
+
+use super::util::{get_host_path, get_virt_drive_name};
+pub type ArcMutexBoxDevice = Arc<Mutex<Box<dyn Device>>>;
+
+/// block_index and released_block_index are used to search an available block index
+/// in Sandbox.
+///
+/// @block_index generally default is 1 for <vdb>;
+/// @released_block_index for blk devices removed and indexes will released at the same time.
+#[derive(Clone, Debug, Default)]
+struct SharedInfo {
+    block_index: u64,
+    released_block_index: Vec<u64>,
+}
+
+impl SharedInfo {
+    fn new(index: u64) -> Self {
+        SharedInfo {
+            block_index: index,
+            released_block_index: vec![],
+        }
+    }
+
+    // declare the available block index
+    fn declare_device_index(&mut self) -> Result<u64> {
+        let current_index = if let Some(index) = self.released_block_index.pop() {
+            index
+        } else {
+            self.block_index
+        };
+        self.block_index += 1;
+
+        Ok(current_index)
+    }
+
+    fn release_device_index(&mut self, index: u64) {
+        self.released_block_index.push(index);
+        self.released_block_index.sort_by(|a, b| b.cmp(a));
+    }
+}
+
+// Device manager will manage the lifecycle of sandbox device
+pub struct DeviceManager {
+    devices: HashMap<String, ArcMutexBoxDevice>,
+    hypervisor: Arc<dyn Hypervisor>,
+    shared_info: SharedInfo,
+}
+
+impl DeviceManager {
+    pub fn new(hypervisor: Arc<dyn Hypervisor>) -> Result<Self> {
+        let devices = HashMap::<String, ArcMutexBoxDevice>::new();
+        Ok(DeviceManager {
+            devices,
+            hypervisor,
+            shared_info: SharedInfo::new(1),
```

Comment:
Why should pass the `1` to the `SharedInfo`, I think it should be hidden in the `SharedInfo::new()`.

Reply:
fixed